samples/bluetooth/channel_sounding/src/connected_cs.c

```diff
@@ -0,0 +1,498 @@
+			return 0;
+		}
+
+		k_sem_take(&sem_remote_capabilities_obtained, K_FOREVER);
+
+		struct bt_le_cs_create_config_params config_params = {
+			.id = CS_CONFIG_ID,
+			.main_mode_type = BT_CONN_LE_CS_MAIN_MODE_2,
+			.sub_mode_type = BT_CONN_LE_CS_SUB_MODE_1,
+			.min_main_mode_steps = 2,
+			.max_main_mode_steps = 10,
+			.main_mode_repetition = 0,
+			.mode_0_steps = NUM_MODE_0_STEPS,
+			.role = role_selection,
+			.rtt_type = BT_CONN_LE_CS_RTT_TYPE_AA_ONLY,
+			.cs_sync_phy = BT_CONN_LE_CS_SYNC_1M_PHY,
+			.channel_map_repetition = 1,
+			.channel_selection_type = BT_CONN_LE_CS_CHSEL_TYPE_3B,
+			.ch3c_shape = BT_CONN_LE_CS_CH3C_SHAPE_HAT,
+			.ch3c_jump = 2,
+		};
+
+		bt_le_cs_set_valid_chmap_bits(config_params.channel_map);
+
+		err = bt_le_cs_create_config(connection, &config_params,
+					     BT_LE_CS_CREATE_CONFIG_CONTEXT_LOCAL_AND_REMOTE);
+		if (err) {
+			printk("Failed to create CS config (err %d)\n", err);
+			return 0;
+		}
+
+		k_sem_take(&sem_config_created, K_FOREVER);
+
+		err = bt_le_cs_security_enable(connection);
+		if (err) {
+			printk("Failed to start CS Security (err %d)\n", err);
+			return 0;
+		}
+
+		k_sem_take(&sem_cs_security_enabled, K_FOREVER);
+
+		const struct bt_le_cs_set_procedure_parameters_param procedure_params = {
+			.config_id = CS_CONFIG_ID,
+			.max_procedure_len = 12,
+			.min_procedure_interval = 100,
+			.max_procedure_interval = 100,
+			.max_procedure_count = 0,
+			.min_subevent_len = 6750,
+			.max_subevent_len = 6750,
+			.tone_antenna_config_selection =
+				BT_LE_CS_TONE_ANTENNA_CONFIGURATION_INDEX_ONE,
+			.phy = 0x01,
+			.tx_power_delta = 0x80,
+			.preferred_peer_antenna = 1,
+			.snr_control_initiator = BT_LE_CS_INITIATOR_SNR_CONTROL_NOT_USED,
+			.snr_control_reflector = BT_LE_CS_REFLECTOR_SNR_CONTROL_NOT_USED,
+		};
+
+		err = bt_le_cs_set_procedure_parameters(connection, &procedure_params);
+		if (err) {
+			printk("Failed to set procedure parameters (err %d)\n", err);
+			return 0;
+		}
+
+		struct bt_le_cs_procedure_enable_param params = {
+			.config_id = CS_CONFIG_ID,
+			.enable = 1,
+		};
+
+		err = bt_le_cs_procedure_enable(connection, &params);
+		if (err) {
+			printk("Failed to enable CS procedures (err %d)\n", err);
+			return 0;
+		}
+	}
+
+	while (true) {
+		k_sem_take(&sem_procedure_done, K_FOREVER);
+
+		if (role_selection == BT_CONN_LE_CS_ROLE_REFLECTOR) {
+			write_params.func = write_func;
+			write_params.handle = step_data_attr_handle;
+			write_params.length = STEP_DATA_BUF_LEN;
+			write_params.data = &latest_local_steps[0];
+			write_params.offset = 0;
+
+			err = bt_gatt_write(connection, &write_params);
+			if (err) {
+				printk("Write failed (err %d)\n", err);
+				return 0;
+			}
+
+			err = k_sem_take(&sem_written, K_SECONDS(10));
+			if (err) {
+				printk("Timed out during GATT write\n");
+				return 0;
+			}
+		}
+
+		if (role_selection == BT_CONN_LE_CS_ROLE_INITIATOR) {
+			k_sem_take(&sem_data_received, K_FOREVER);
+
+			estimate_distance(
```

Comment:
Did you measure how much time the processing in `estimate_distance` takes for the default configuration used in the sample? Just curious since I see it uses some floating point

Reply:
Using k_uptime_ticks() to record how much time is spent in this function I was able to estimate that it takes about 2.5-3ms to run estimate_distance() (excluding the  printk() calls at the end). This is with a bit less than 500 bytes of mostly mode-2 steps covering about 35-40 channels. So I suppose we can extrapolate that running this code for all channels would take ~5-6ms.

I'm not sure if this is considered fast or slow, I guess it depends on how fast the devices are moving. What do you think?

Reply:
Using single-precision floats and CONFIG_FPU=y CONFIG_FPU_SHARING=y I'm able to reduce this by about 40%

Reply:
I think 5-6 ms is good enough for this sample, thanks for checking